app/models/external_activity.rb

```diff
@@ -109,6 +109,13 @@ class ExternalActivity < ActiveRecord::Base
   include SearchModelInterface
   include Archiveable
 
+  #
+  # Override the material_type method from SearchModelInterface
+  #
+  def material_type
+    template_type ? template_type : attributes['material_type']
+  end
+
```

Comment:
> I think this approach will be confusing: a method name that matches a database field but it doesn't always return the database field's value.

It is a bit confusing. And this is why I had to go into this `attributes['material_type']` approach after trying to figure out why I was not seeing the attribute as I expected. (The SearchMdoelInteface also defines this `material_type` method to be something else entirely.) 

Do we think this `before_validation` approach is less confusing?

Question:
> When there is a template_type of 'Investigation' the material_type in the database will be 'Activity'.

Why would this be? This won't be the case for any existing ExternalActivities that get handled by the db migration. If they use the publish_sequence() API this should also be set correctly.

I suppose code could directly create an ExternalActivity without setting material_type appropriately. And so the above `before_validation` addresses this case?

Reply:
There appears to be several tests that create ExternalActivities with templates.  These tests will fail without either your code above, or the before_validation code.

You are right that having the code in publish_sequence will prevent a mismatch from happening in the runtime. But as these comments show this has already confused me, I think it is because there are two places where this is being handled. If we switch to the before_validation approach, then we should be able to remove the code in publish_sequence.  So now it will all be handled in one place.

Reply:
Should this be `:if => :template_changed?` ?

Reply:
This test failure can be addressed by ensuring that changes to the template (as well as the template_type) update the material_type.
So this would work:

```
+  #
+  # Ensure changes to the template update the material_type
+  #
+  alias_method :original_template=, :template=
+  def template=(t)
+    original_template   = t
+    self.material_type  = t.class.name
   end

```

But this breaks search result specs which are not expecting templates?

```
      describe "template items should not be included in results by default" do
        let(:template_ivs)  { collection(:investigation_template, 2, public_opts) }
        let(:template_acts) { collection(:activity_template, 2, public_opts) }
        let(:materials)     { [public_items, template_ivs, template_acts].flatten }
        it "results should not include any of the template activities" do
          template_acts.each do |act|
            subject.results[:all].should_not include act
          end
        end

```

Reply:
Yes the spec 'should return template_type for EAs with templates'  was failing because save was not being called.  Unless save is called the validations won't be called. So this illustrates an interesting problem with the callback approach. If some code (like this spec) is assuming the material type to be updated immediately then you will need to override the `template=` method. 

I think it would be OK to change this spec to call save first, so you don't need to override the `template=` method

In your overriding method you might need to call `self.original_template = t` instead of just `original_template = t`.  I always forget when self is needed. That might be why the 'template items should not be included in results by default' spec is breaking. But I don't think so.

The 'template items should not be included in results by default' spec is trying to make sure the search results don't include the hidden activity and investigations that are backing the ExternalActivities.  The spec is using this factory to make some of these templates:
```
Factory.define :investigation_template, parent: :investigation do |f|
  f.after_create do |inv|
    FactoryGirl.create_list(:external_activity, 2, template: inv, url: "http://activity.external.com/1/2/3")
  end
end
```

So the spec should essentially be doing this:
```
inv = Investigation.create(public_opts)
ExternalActivity.create(template: inv, url: "http://activity.external.com/1/2/3")
ExternalActivity.create(template: inv, url: "http://activity.external.com/1/2/3")
```

And then it is checking that the search results don't include these templates.  This implies that Solr is not correctly updating the the template Activity and Investigation objects. 
Perhaps that is broken by the before_validation code, or perhaps it is broken by the overriding the `template=` method.  I would not expect `template=` to be called when `create(template: inv)` is called. So I'd guess there is something wrong with before_validation code. If you take these changes out does that spec pass? I think it must because all the tests are passing on this PR currently.

Reply:
> In your overriding method you might need to call self.original_template = t

Well spotted. This is the problem (which uses in turn is_template from the sunspot searchable interface to do an sql query against the act, inv and ea tables.)

Reply:
Hmm. So I guess the Factory.create methods are actually calling the `field=` methods.

Reply:
I've reverted the `save!` call from the spec which modified the `template` and I've fixes the alias_method to use `self` when calling the original setter. This should have both failed specs passing now.